### PR TITLE
[WIP] Framework: add separate entry point for 'login'

### PR DESCRIPTION
**Pull request by flootr**
**Don't merge** ⚠️ 

Explorations if we can get a separate login entry point working. The goal is to get rid of anything which is not needed for login and serve a way faster experience and a smaller bundle.

### CODE REVIEW

File: client/boot/login.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import loginModule from 'login';
+import createReduxStoreFromPersistedInitialState from 'state/initial-state';
+import * as controller from 'controller/index.web';
+import { setupContextMiddleware } from './common';
+
+const debug = debugFactory( 'calypso' );
+
+window.AppBoot = () => {
+	debug( 'boot login page' );
+	createReduxStoreFromPersistedInitialState( reduxStore => {
```
Comment (blowery) [inline client/boot/login.js]: we'll want a different version of this just for login with a different reducer tree and probably no persistence support

**Comment by jsnajdr** [comment]
I did some investigation about why the `login` chunk is still so huge -- only few percent smaller than `build` -- and why it needs a few huge shared chunks like `vendor~build~login`. I think it's related to how tree shaking is done. Here is what I think is going on:

How Webpack does shared modules and tree shaking:
1. Builds a module dependency tree for all entry points and async-loaded chunks
2. Identifies some common subtrees, measures their size (before any tree shaking or other processing) and creates a set of shared chunks based on some optimization algorithm.
3. Shakes the tree inside the chunks (whose contents are already determined) to remove unused code.

Shared chunks are imported by multiple other chunks and each chunk can import different bindings. For example:

```js
// login.js
import { isDomainConnectAuthorizePath } from 'lib/domains/utils';
```

```js
// checkout.js
import { getDomainNameFromReceiptOrCart } from 'lib/domains/utils';
```

Now Webpack decides that `lib/domains/utils` is a good candidate to bundle in a chunk shared between `login` and `checkout`. What could possibly go wrong?

- `isDomainConnectAuthorizePath` is a trivial one-line function.
- `getDomainNameFromReceiptOrCart` is a monster that depends on many other libraries, e.g., `lib/cart-values` or `lib/products-values`

When tree shaking the shared chunk, Webpack cannot remove `getDomainNameFromReceiptOrCart` and its dependencies, because it's used at least by one user!

Suddenly a lot of unneeded code is loaded by the `login` chunk. It was not removed by tree shaking, because Webpack shakes the shared chunk as a whole, it doesn't do one custom version of the `lib/domains/utils` library just for `login` and another one just for `checkout`.

A version of `lib/domains/utils` that's custom-built just for `login` would be just a few bytes of code, but that doesn't happen.

That's my theory about why our `login` chunk is so big. I'm not sure if it's correct and I don't have any solution either 🙃

**Comment by alisterscott** [comment]
This one needs a rebase. Are we continuing with it?

**Comment by flootr** [comment]
@alisterscott I'm going to rebase and continue working on this soon.

**Comment by flootr** [comment]
this became stale, closing for now

**Pull request closed without merging**